[docs] add repack doc (#44883)
expo/expo: docs/pages/build-reference/repack.mdx

---
title: Repack app

> description: Repackage an existing APK, IPA, or .app with an updated JS bundle without a full native rebuild.

Reviewer: ```suggestion
description: Repackage an existing APK, IPA, or .app with an updated JavaScript bundle without a full native rebuild.
```

Let's use the full form here since the description field of a page directly affects to SEO of this page.

---

import { Terminal } from '~/ui/components/Snippet';

> [`@expo/repack-app`](https://www.npmjs.com/package/@expo/repack-app) is a CLI tool that repackages an existing Android APK, iOS IPA, or iOS **.app** bundle with an updated JavaScript bundle, assets, and app metadata — such as the app name, version, and bundle identifier — without performing a full native rebuild. It also handles Expo-specific concerns like the expo-updates manifest and embedded assets. Because no native compilation is involved, a repack is typically much faster than a full native build.

Reviewer: ```suggestion
[`@expo/repack-app`](https://www.npmjs.com/package/@expo/repack-app) is a CLI tool that repackages an existing Android APK, iOS IPA, or iOS **.app** bundle without performing a full native build. It updates the existing artifact with a new JavaScript (JS) bundle, assets, and app metadata (such as, app name, version, and package name or bundle identifier). Since there is no native compilation involved, a repack is typically much faster than a full native build.
```

I think the first sentence is a bit hard to read because of its length and mixing what it does and how it does it.

> To learn more about how fingerprint and repack work together to accelerate CI, see [Accelerating Continuous Integration with Fingerprint and Repack in EAS Workflows](https://expo.dev/blog/accelerating-continuous-integration-with-fingerprint-repack-in-eas-workflows).

Reviewer: ```suggestion
To learn more about how fingerprint and repack work together to accelerate CI, see [Accelerating Continuous Integration with Fingerprint and Repack in EAS Workflows](https://expo.dev/blog/accelerating-continuous-integration-with-fingerprint-repack-in-eas-workflows) blog post.
```

## Prerequisites

> - An existing build artifact produced from your Expo project: an APK for Android; an IPA or **.app** bundle for iOS.

Reviewer: ```suggestion
- An existing build artifact produced from your Expo project: an APK for Android, or an IPA or **.app** bundle for iOS.
```

> - (Optional) Signing credentials if the repacked artifact needs to be installable on a device — see [Signing](#signing).

Reviewer: ```suggestion
- (Optional) Signing credentials if the repacked artifact needs to be installable on a device. For more information, see [Signing](#signing).
```

## When to use repack

> Repack assumes the source binary's native side is unchanged since it was built. If you added a native dependency, changed a config plugin, or upgraded the Expo SDK, a repacked artifact will have JS that expects native APIs that do not exist, and will likely crash at runtime. Use [fingerprint](/versions/latest/sdk/fingerprint/) to compare the source binary's native identity against your current project — if the fingerprints match, repack is safe; if they differ, do a full native build instead.

Reviewer: ```suggestion
Repack assumes the source binary's native side is unchanged since it was built. If you added a native dependency, changed a [config plugin](/config-plugins/introduction/), or upgraded the Expo SDK, a repacked artifact will have JS that expects native APIs that do not exist, and will likely crash at runtime. Use [fingerprint](/versions/latest/sdk/fingerprint/) to compare the source binary's native identity against your current project. If the fingerprints match, then repack is safe. If they differ, do a full native build instead.
```

> Repack is not a replacement for [EAS Update](/eas-update/introduction/). EAS Update ships a new JS bundle to apps that are already installed, and users see it on the next launch. Repack produces a new installable artifact. A good rule of thumb: use repack for internal testing (QA devices, testers, CI smoke tests) and use EAS Update to deliver JS changes to production users.

Reviewer: ```suggestion
Repack is not a replacement for [EAS Update](/eas-update/introduction/). EAS Update ships a new JS bundle to apps that are already installed, and users see it on the next launch. Repack produces a new installable artifact. A good rule of thumb is to use repack for internal testing (QA devices, testers, CI smoke tests) and use EAS Update to deliver JS changes to production users.
```

> ## Standalone CLI usage

Reviewer: I think there should be an H2 for a Usage section here that could briefly mentions that there are two ways to use repack. 

And both "Standalone CLI usage" (Standalone CLI) and "Usage in EAS Workflows" (EAS Workflows) will be subsections as H3s to the Usage section.

This will help improving this doc's structure and also allow a reader to quickly scan it.

> Run the command from your project's root directory. At minimum, `--platform` and `--source-app` are required. The tool runs `npx expo export:embed` under the hood to produce a fresh JS bundle, swaps it into the source binary, and writes the output artifact.

Reviewer: ```suggestion
You need to run the commands specified for Android and iOS platforms below from your project's root directory. At minimum, `--platform` and `--source-app` are required. The tool runs `npx expo export:embed` under the hood to produce a fresh JS bundle, swaps it into the source binary, and writes the output artifact.
```

> The output format always matches the `--source-app` input: an APK input produces an APK, an IPA produces an IPA, and a **.app** bundle produces a **.app** bundle.

Reviewer: ```suggestion
The output format always matches the `--source-app` input. An APK input produces an APK, an IPA produces an IPA, and a **.app** bundle produces a **.app** bundle.
```

### Android

<Terminal cmd={['$ npx @expo/repack-app --platform android --source-app MyApp.apk']} />

### iOS

Using an IPA:

<Terminal cmd={['$ npx @expo/repack-app --platform ios --source-app MyApp.ipa']} />

> Using a **.app** bundle (for simulator builds — iOS device artifacts are always IPAs):

Reviewer: ```suggestion
Using a **.app** bundle (for simulator builds, the iOS device artifacts are always IPAs):
```

> <Terminal cmd={['$ npx @expo/repack-app --platform ios --source-app MyApp.app']} />

Reviewer: These two sections should be Tabs (Android, iOS). This will help improve the readability (make the page a bit quicker to scan) of the page since both of these sections only display commands to run and the explanation is cover above.

## Signing

To produce an artifact installable on a physical device, pass signing credentials alongside `--platform` and `--source-app`. Without them, the Android APK is output unsigned and the iOS IPA cannot be installed on a device. iOS **.app** bundles intended for the simulator do not need signing and can skip this step.

If you do not already have signing credentials locally, see [App credentials](/app-signing/app-credentials/) for how to obtain a keystore (Android) or a signing identity and provisioning profile (iOS).

> ### Android

Reviewer: My suggestion here would be to use Tabs for Android and iOS instead of H3s (same as above).

Re-sign the repacked APK with a keystore:

<Terminal
  cmd={[
    '$ npx @expo/repack-app \\',
    '  --platform android \\',
    '  --source-app MyApp.apk \\',

>     '  --ks keystore.jks \\',
>     '  --ks-key-alias my-alias',
>   ]}
>   cmdCopy="npx @expo/repack-app --platform android --source-app MyApp.apk --ks keystore.jks --ks-key-alias my-alias"
> />

Reviewer: ```suggestion
<Terminal
  cmd={[ '$ npx @expo/repack-app --platform android --source-app MyApp.apk --ks keystore.jks --ks-key-alias my-alias' ]}
/>
```

I don't think we should break the command with newline characters here unless there is a reason (that I might be unaware). We have other docs where we similarly lengthy commands and readers usually use the "Copy" button.

See the [Android-specific options](#android-specific-options) for keystore password flags and other settings.

### iOS

Pass a signing identity and provisioning profile:

> **Note**: iOS repacking supports ad-hoc and development signing only.

<Terminal
  cmd={[
    '$ npx @expo/repack-app \\',
    '  --platform ios \\',
    '  --source-app MyApp.ipa \\',

>     '  --signing-identity "Apple Distribution: ..." \\',
>     '  --provisioning-profile /path/to/profile.mobileprovision',
>   ]}
>   cmdCopy='npx @expo/repack-app --platform ios --source-app MyApp.ipa --signing-identity "Apple Distribution: ..." --provisioning-profile /path/to/profile.mobileprovision'
> />

Reviewer: ```suggestion
<Terminal
  cmd={[ '$ npx @expo/repack-app --platform ios --source-app MyApp.ipa --signing-identity "Apple Distribution: ..." --provisioning-profile /path/to/profile.mobileprovision',
/>
```

See the [iOS-specific options](#ios-specific-options) for the full list of signing flags.

> ## Use cases

Reviewer: This such a good and concise section. I think if it could be moved just after "When to use repack", that will be good structurally for this documentation page.

- **QA cycles**: Distribute one base build to testers, then repack with JS fixes for each iteration without waiting for a native rebuild.
- **CI optimization**: Build native once per fingerprint, then repack the JS bundle on every subsequent PR to cut wait times from minutes to seconds.
- **Branch testing**: Test multiple JS branches against the same native binary to isolate JavaScript-only changes.

## Usage in EAS Workflows

EAS Workflows provides a pre-packaged `repack` job type that handles signing and build management automatically. See [EAS Workflows pre-packaged jobs](/eas/workflows/pre-packaged-jobs/#repack) for full syntax, parameters, and examples.

## `--js-bundle-only` mode

> By default, repack updates the JS bundle, assets, and native metadata (app name, version, bundle identifier, and the expo-updates manifest). Pass `--js-bundle-only` when you intentionally want to update only the JS bundle and leave all native config untouched.

Reviewer: ```suggestion
By default, repack updates the JS bundle, assets, and app metadata (app name, version, bundle identifier, and the expo-updates manifest). Pass `--js-bundle-only` when you intentionally want to update only the JS bundle and leave all native config untouched.
```

I think we should be consistent with the terms and decide between using either "app metadata" or "native metadata".

<Terminal
  cmd={[
    '$ npx @expo/repack-app \\',
    '  --platform android \\',
    '  --source-app MyApp.apk \\',
    '  --js-bundle-only',

>   ]}
>   cmdCopy="npx @expo/repack-app --platform android --source-app MyApp.apk --js-bundle-only"
> />

Reviewer: ```suggestion
<Terminal cmd={['$ npx @expo/repack-app --platform android --source-app MyApp.apk --js-bundle-only']} />
```

## Limitations

- Repack is not recommended for production App Store or Play Store submissions. Production builds should go through the complete build pipeline for correct symbolication and signing.

## CLI reference

```text
Usage: @expo/repack-app [options] [project-root]
```

### Arguments

| Argument         | Description                                                          |
| ---------------- | -------------------------------------------------------------------- |
| `[project-root]` | Path to the project root. Defaults to the current working directory. |

### Options

| Option                                    | Description                                                                                                                                              |
| ----------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `-p, --platform <platform>`               | **Required.** Platform to repack the app for (`android` or `ios`).                                                                                       |

> | `--source-app <path>`                     | **Required.** Path to the source app (APK for Android; IPA or **.app** for iOS). The output format matches the input.                                    |

Reviewer: ```suggestion
| `--source-app <path>`                     | **Required.** Path to the source app (APK for Android, or IPA or **.app** for iOS). The output format matches the input.                                    |
```

<Tabs>

<Tab label="Android">

<Terminal
  cmd={[
    '$ npx @expo/repack-app --platform android --source-app MyApp.apk --ks keystore.jks --ks-key-alias my-alias',
  ]}
/>

See the [Android-specific options](#android-specific-options) for keystore password flags and other settings.

</Tab>

<Tab label="iOS">

Pass a signing identity and provisioning profile:

> **Note**: iOS repacking supports ad-hoc and development signing only.

<Terminal
  cmd={[
    '$ npx @expo/repack-app --platform ios --source-app MyApp.ipa --signing-identity "Apple Distribution: ..." --provisioning-profile /path/to/profile.mobileprovision',
  ]}
/>

See the [iOS-specific options](#ios-specific-options) for the full list of signing flags.

</Tab>

</Tabs>

## `--js-bundle-only` mode

By default, repack updates the JS bundle, assets, and app metadata (app name, version, bundle identifier, and the expo-updates manifest). Pass `--js-bundle-only` when you intentionally want to update only the JS bundle and leave all native config untouched.

<Terminal
  cmd={['$ npx @expo/repack-app --platform android --source-app MyApp.apk --js-bundle-only']}
/>

## Limitations

> - Repack is not recommended for production App Store or Play Store submissions. Production builds should go through the complete build pipeline for correct symbolication and signing.

Reviewer: ```suggestion
Repack is not recommended for production Google Play Store or Apple App Store submissions. Production builds should go through the complete build pipeline for correct symbolication and signing.
```

- Let's remove the item list since there is only one item right now.
- Also, fix the platform order and app store names.
